pkg/storage/mvcc.go

```diff
@@ -4256,6 +4301,221 @@ func MVCCGarbageCollect(
 	return nil
 }
 
+// MVCCGarbageCollectRanges is similar in functionality to MVCCGarbageCollect but
+// operates on range keys. It does sanity checks that no values exist below
+// range tombstones so that no values are exposed in case point values GC was
+// not performed correctly by level above.
+// Note that method requires start and end keys for the range since we don't
+// want underlying iterators to trim exposed range fragments and we don't
+// otherwise have a simple means to extend ranges beyond provided range keys.
+func MVCCGarbageCollectRanges(
+	ctx context.Context,
+	rw ReadWriter,
+	rangeStartKey roachpb.Key,
+	rangeEndKey roachpb.Key,
+	ms *enginepb.MVCCStats,
+	rangeKeys []roachpb.GCRequest_GCRangeKey,
+	timestamp hlc.Timestamp,
+) error {
+
+	var count int64
+	defer func(begin time.Time) {
+		log.Eventf(ctx,
+			"done with GC evaluation for %d range keys at %.2f keys/sec. Deleted %d entries",
+			len(rangeKeys), float64(len(rangeKeys))*1e9/float64(timeutil.Since(begin)), count)
+	}(timeutil.Now())
+
+	// If there are no keys then there is no work.
+	if len(rangeKeys) == 0 {
+		return nil
+	}
+
+	// Validate range keys are well formed.
+	highestTS := hlc.Timestamp{}
+	rks := make([]MVCCRangeKey, len(rangeKeys))
+	for i, rk := range rangeKeys {
+		rks[i] = MVCCRangeKey{StartKey: rk.StartKey, EndKey: rk.EndKey, Timestamp: rk.Timestamp}
+		if err := rks[i].Validate(); err != nil {
+			return errors.Wrap(err, "failed to validate gc range keys in mvcc gc")
+		}
+		if highestTS.Less(rk.Timestamp) {
+			highestTS = rk.Timestamp
+		}
+	}
+
+	sort.Slice(rks, func(i, j int) bool {
+		return rks[i].Compare(rks[j]) < 0
+	})
+
+	// Validate that keys are non-overlapping.
+	var lastEndKey roachpb.Key
+	for i, rangeKey := range rks {
+		if rangeKey.StartKey.Compare(lastEndKey) < 0 {
+			return errors.Errorf("range keys in gc request should be non-overlapping: %s vs %s",
+				rks[i-1].String(), rangeKey.String())
+		}
+		lastEndKey = rangeKey.EndKey
+	}
+
+	// Bound the iterator appropriately for the set of keys we'll be garbage
+	// collecting.
+	iter := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
```

Comment:
`IterKeyTypeRangesOnly` can't surface intents by definition, so we can avoid the overhead of the `intentInterleavingIterator` by using `MVCCKeyIterKind`.